src/Libraries/Microsoft.Extensions.DataIngestion/Chunkers/SemanticSimilarityChunker.cs

```diff
@@ -0,0 +1,146 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Numerics.Tensors;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.AI;
+using Microsoft.Shared.Diagnostics;
+
+namespace Microsoft.Extensions.DataIngestion.Chunkers;
+
+/// <summary>
+/// Splits a <see cref="IngestionDocument"/> into chunks based on semantic similarity between its elements.
+/// </summary>
+public sealed class SemanticSimilarityChunker : IngestionChunker<string>
+{
+    private readonly ElementsChunker _elementsChunker;
+    private readonly IEmbeddingGenerator<string, Embedding<float>> _embeddingGenerator;
+    private readonly float _thresholdPercentile;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="SemanticSimilarityChunker"/> class.
+    /// </summary>
+    /// <param name="embeddingGenerator">Embedding generator.</param>
+    /// <param name="options">The options for the chunker.</param>
+    /// <param name="thresholdPercentile">Threshold percentile to consider the chunks to be sufficiently similar.</param>
+    public SemanticSimilarityChunker(
+        IEmbeddingGenerator<string, Embedding<float>> embeddingGenerator,
+        IngestionChunkerOptions options,
+        float thresholdPercentile = 95.0f)
+    {
+        _embeddingGenerator = Throw.IfNull(embeddingGenerator);
+        _elementsChunker = new(options);
+
+        if (thresholdPercentile < 0f || thresholdPercentile > 100f)
+        {
+            Throw.ArgumentOutOfRangeException(nameof(thresholdPercentile), "Threshold percentile must be between 0 and 100.");
+        }
+
+        _thresholdPercentile = thresholdPercentile;
+    }
+
+    /// <inheritdoc/>
+    public override async IAsyncEnumerable<IngestionChunk<string>> ProcessAsync(IngestionDocument document,
+        [EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        cancellationToken.ThrowIfCancellationRequested();
+
+        _ = Throw.IfNull(document);
+
+        if (document.Sections.Count == 0)
+        {
+            yield break;
+        }
+
```

Comment:
such documents should be very rare, let's not optimize for them

```suggestion
        if (document.Sections.Count == 0)
        {
            yield break;
        }

```